### PR TITLE
[Serializer] XmlEncoder: Make load flags configurable

### PR DESCRIPTION
**Pull request by dunglas**
| Q | A |
| --- | --- |
| Branch | master |
| Bug fix? | no |
| New feature? | yes |
| BC breaks? | no |
| Deprecations? | no |
| Tests pass? | yes |
| Fixed tickets | #17987 |
| License | MIT |
| Doc PR | todo |

### CODE REVIEW

File: src/Symfony/Component/Serializer/Encoder/XmlEncoder.php [hunk 1]
```diff
@@ -30,15 +30,18 @@ class XmlEncoder extends SerializerAwareEncoder implements EncoderInterface, Dec
...
      *
-     * @param string $rootNodeName
+     * @param string   $rootNodeName
+     * @param int|null $loadOptions
      */
-    public function __construct($rootNodeName = 'response')
+    public function __construct($rootNodeName = 'response', $loadOptions = null)
     {
         $this->rootNodeName = $rootNodeName;
+        $this->loadOptions = $loadOptions ?: LIBXML_NONET | LIBXML_NOBLANKS;
```
Comment (xabbuh) [inline src/Symfony/Component/Serializer/Encoder/XmlEncoder.php]: `$this->loadOptions = null !== $loadOptions ?: LIBXML_NONET | LIBXML_NOBLANKS;`

Comment (dunglas) [inline src/Symfony/Component/Serializer/Encoder/XmlEncoder.php]: done

Comment (xabbuh) [inline src/Symfony/Component/Serializer/Encoder/XmlEncoder.php]: oh wait, the suggestion was bad, it must now be `$this->loadOptions = null !== $loadOptions ? $loadOptions : LIBXML_NONET | LIBXML_NOBLANKS;`

Comment (dunglas) [inline src/Symfony/Component/Serializer/Encoder/XmlEncoder.php]: I copy/pasted without care, sorry. It's fixed now.

File: src/Symfony/Component/Serializer/Encoder/XmlEncoder.php [hunk 2]
```diff
@@ -80,8 +83,10 @@ public function decode($data, $format, array $context = array())
         $disableEntities = libxml_disable_entity_loader(true);
         libxml_clear_errors();
 
+        $loadOptions = isset($context['xml_load_options']) ? $context['xml_load_options'] : $this->loadOptions;
```
Comment (xabbuh) [inline src/Symfony/Component/Serializer/Encoder/XmlEncoder.php]: Is it really a good idea to make this overridable through the context?

Comment (dunglas) [inline src/Symfony/Component/Serializer/Encoder/XmlEncoder.php]: I think you're right, I'll update the PR.

**Comment by dunglas** [comment]
ping @symfony/deciders

File: src/Symfony/Component/Serializer/Encoder/XmlEncoder.php [hunk 3]
```diff
@@ -30,15 +30,18 @@ class XmlEncoder extends SerializerAwareEncoder implements EncoderInterface, Dec
...
     private $context;
     private $rootNodeName = 'response';
+    private $loadOptions;
 
     /**
      * Construct new XmlEncoder and allow to change the root node element name.
      *
-     * @param string $rootNodeName
+     * @param string   $rootNodeName
+     * @param int|null $loadOptions
```
Comment (xabbuh) [inline src/Symfony/Component/Serializer/Encoder/XmlEncoder.php]: We should add a short explanation what `$loadOptions` means (something like "a bit field of LIBXML_ constants").

**Comment by xabbuh** [comment]
:+1: (apart from my last minor comment)

**Comment by dunglas** [comment]
@xabbuh comment added

**Comment by fabpot** [comment]
Thank you @dunglas.

**Comment by rvanlaak** [comment]
Great :+1: Can someone point me in the direction how to override this settings for `FOS\RestBundle\EventListener\BodyListener:114` as explained in https://github.com/symfony/symfony/issues/17987#issue-137933421?

**Comment by dunglas** [comment]
Add this class with the new options as a new encoder: http://symfony.com/doc/current/cookbook/serializer.html#adding-normalizers-and-encoders

**Pull request closed without merging**